scripts/fetch_articles.py

```diff
@@ -1,181 +1,313 @@
+    article_basics = dict()
+    # record article dataset relationship, keys are article id
+    article_datasets: Dict[str, Set] = dict()
+    # for progress purpose
+    dataset_count = 0
+    # one dataset could have multiple articles, keys are dataset id, same naming pattern for other record type
+    article_for_datasets: Dict[str, Set] = dict()
+
+    # for all datasets existing in the Elasticsearch, search for the publications based on the dataset accession
+    for dataset_id in datasets.keys():
+        # logging progress, not related to the main algorithm
+        dataset_count = dataset_count + 1
+        if dataset_count % 200 == 0:
+            logger.info(f'Processed {dataset_count} datasets')
+        # get dataset related publication using europe PMC search API
+        url = f"https://www.ebi.ac.uk/europepmc/webservices/rest/search?query={dataset_id}&format=json"
+        epmc_result = requests.get(url).json()
+        hit_count = epmc_result['hitCount']
+        if hit_count != 0:
+            epmc_hits = epmc_result['resultList']['result']
+            for hit in epmc_hits:
+                # ignore preprints determined by two fields pubType and source
+                if 'pubType' in hit and hit['pubType'] == 'preprint':
+                    continue
+                if 'source' in hit and hit['source'] == 'PPR':
+                    continue
+                # determine the article id which will be used in ES, PMC id is preferred, because
+                # 1) it has PMC prefix rather than a string of digits
+                # 2) PMC guarantees open access, more likely to have dataset accession linked
+                article_id = determine_article_id(hit)
+                if len(article_id) == 0:
+                    logger.error(f"Study {dataset_id} has related article without Identifier")
+                    continue
+                # new article
+                if article_id not in article_details:
+                    es_article = dict()
+                    es_article = parse_field(es_article, hit, 'pmcId', 'pmcid')
+                    es_article = parse_field(es_article, hit, 'pubmedId', 'pmid')
+                    es_article = parse_field(es_article, hit, 'doi', 'doi')
+                    es_article = parse_field(es_article, hit, 'title', 'title')
+                    es_article = parse_field(es_article, hit, 'authorString', 'authorString')
+                    es_article = parse_field(es_article, hit, 'journal', 'journalTitle')
+                    es_article = parse_field(es_article, hit, 'issue', 'issue')
+                    es_article = parse_field(es_article, hit, 'volume', 'journalVolume')
+                    es_article = parse_field(es_article, hit, 'year', 'pubYear')
+                    es_article = parse_field(es_article, hit, 'pages', 'pageInfo')
+                    es_article = parse_field(es_article, hit, 'isOpenAccess', 'isOpenAccess')
```

Comment:
This could be simplified. Create dict with keys as third value in parse_field function and values as fourth value in in parse_field function. For example dict = {'pmcId': 'pmcid', 'pubmedId': 'pmid'}. Then you can get through the dict and use key, value in function:
for k,v in dict.items():
    es_article = parse_field(es_article, hit, k, v)

Reply:
done in a22c009